tsl/src/continuous_aggs/refresh.c

```diff
@@ -1040,6 +1040,12 @@ continuous_agg_split_refresh_window(ContinuousAgg *cagg, InternalTimeRange *orig
 			compute_inscribed_bucketed_refresh_window(cagg, &refresh_window, bucket_width);
 	}
 
+	/* We must cap the refresh window at the invalidation threshold to don't produce unnecessary
+	 * batches */
+	int64 invalidation_threshold =
+		invalidation_threshold_get(cagg->data.raw_hypertable_id, refresh_window.type);
+	refresh_window.end = MIN(invalidation_threshold, refresh_window.end);
```

Comment:
Don't you need to cap the refresh window at the beginning as well? In theory, the refresh window could be entirely after the threshold, in which case you will cap the end, but not the beginning.

Reply:
Hmmm, I need to think more about it... probably what is necessary here is cap at the MAX(time_dimension) of the hypertable and not the threshold. The idea here is to optimize the batches creation to don't create unnecessary batches.

Reply:
Not sure I followed that, but my point is that if you have a refresh window `10:00-11:00` and you have a threshold at `09:00`, if you just cap the refresh window at the end you will get the range `10:00-09:00`, which is not a valid window at all and could lead to strange behavior later in the code.